### PR TITLE
Refactor launch.json definition

### PR DESCRIPTION
**Pull request by testforstephen**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -147,79 +148,66 @@
...
+                "description": "The program arguments (Optional).",
+                "default": ""
               },
-              "sourcePath": {
+              "vmArguments": {
+                "type": "string",
+                "description": "The VM arguments (Optional).",
+                "default": ""
+              },
+              "classPath": {
```
Comment (akaroml) [inline package.json]: Do we need to use the plural form like "classPaths"?

File: package.json [hunk 2]
```diff
@@ -147,79 +148,66 @@
...
               },
-              "startupClass": {
+              "mainClass": {
                 "type": "string",
-                "description": "Startup class (Should be full qualified name e.g. org.eclipse.xxx.MainClass)",
+                "description": "The fully qualified name of the debuggee application's main class (e.g. org.eclipse.xxx.MainClass).",
                 "default": ""
               },
-              "classpath": {
+              "programArguments": {
```
Comment (akaroml) [inline package.json]: Can you show me an example of how to use this one?

**Review by akaroml — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (testforstephen) [reply package.json]: See the main method of java program `public static void main(String[] args)`.   `programArgument` represents the args array for java program. 
 {
            "type": "java",
            "name": "Java Debug",
            "request": "launch",
            "mainClass": "com.microsoft.debug.HelloWorld",
            "programArguments": "10 11",
            "sourcePath": [
                "${workspaceRoot}"
            ],
            "projectName": "HelloWorld"
        }

File: package.json (hunk 1 above)
Comment (testforstephen) [reply package.json]: Both `java.exe` and eclipse use singular form for class path configuration, so i prefer to use singular form.

**Review by andxu — APPROVED** [review]

File: src/extension.ts [hunk 3]
```diff
@@ -81,26 +81,29 @@ export function activate(context: ExtensionContext) {
...
-										} else if (!config.classpath) {
+										} else if (!config.classPaths || !Array.isArray(config.classPaths) || !config.classPaths.length) {
 											const params: ClasspathResolveRequestParams = {
-												startupClass: config.startupClass,
+												mainClass: config.mainClass,
 												projectName: config.projectName,
 											};
-											const classpath = await languageClient.sendRequest(ClasspathResolveRequest.type, params);
-											config.classpath = classpath;
+											config.classPaths = await languageClient.sendRequest(ClasspathResolveRequest.type, params);
```
Comment (ansyral) [inline src/extension.ts:92]: the return value of the command is string?

Comment (testforstephen) [reply src/extension.ts:92]: i changed it to string array.

File: package.json [hunk 4]
```diff
@@ -138,7 +138,7 @@
     "debuggers": [
       {
         "type": "java",
-        "label": "Java Debug",
+        "label": "Java",
```
Comment (akaroml) [inline package.json:141]: Nice!

File: package.json [hunk 5]
```diff
@@ -148,91 +148,78 @@
...
             "required": [
-              "cwd",
-              "startupClass"
+              "mainClass"
             ],
             "properties": {
-              "cwd": {
+              "projectName": {
+                "type": "string",
+                "description": "Name of the project that contains the main class (Optional, only required if the main class is contained in multiple projects)",
```
Comment (akaroml) [inline package.json]: "only required when the main class resides in multiple projects"

File: package.json [hunk 6]
```diff
@@ -148,91 +148,78 @@
...
+                "default": ""
+              },
+              "mainClass": {
                 "type": "string",
-                "description": "Current working directory (Defaults to workspace root directory).",
-                "default": "${workspaceRoot}"
+                "description": "The main class of the program (Should be fully qualified name, e.g. org.eclipse.xxx.MainClass).",
```
Comment (akaroml) [inline package.json]: "(fully qualified name, e.g. com.xyz.MainClass)"

File: package.json [hunk 7]
```diff
@@ -148,91 +148,78 @@
...
-                "description": "Startup class (Should be full qualified name e.g. org.eclipse.xxx.MainClass)",
+                "description": "The command line arguments passed to the program.",
                 "default": ""
               },
-              "classpath": {
+              "vmArgs": {
                 "type": "string",
-                "default": "",
-                "description": "The classpath to be passed to jvm launcher (Optional, it would override the default classpath)"
+                "description": "The java options and system properties passed to the jvm launcher (e.g. -Xms<size> -Xmx<size> -D<name>=<value>).",
```
Comment (akaroml) [inline package.json]: jvm -> JVM

File: package.json [hunk 8]
```diff
@@ -148,91 +148,78 @@
...
               },
-              "sourcePath": {
+              "classPaths": {
                 "type": "array",
                 "items": {
                   "type": "string"
                 },
-                "default": [],
-                "description": "The source folders of the project."
+                "description": "The classpath passed to the jvm launcher (if not specified, debugger will resolve them automatically from the project configuration).",
```
Comment (akaroml) [inline package.json]: classpath -> classpaths

File: package.json [hunk 9]
```diff
@@ -148,91 +148,78 @@
...
-              "cwd": {
-                "type": "string",
-                "description": "Current working directory (Defaults to workspace root directory).",
-                "default": "${workspaceRoot}"
-              },
               "hostName": {
                 "type": "string",
                 "default": "localhost",
-                "description": "The machine where the debuggee VM was launched on"
+                "description": "The host name or ip address of remote debugging vm."
```
Comment (akaroml) [inline package.json]: debugging vm -> debuggee

File: package.json [hunk 10]
```diff
@@ -148,91 +148,78 @@
...
               },
               "port": {
                 "type": "number",
-                "description": "The debug port that the debuggee VM exposed"
+                "description": "The debug port of remote debugging vm."
```
Comment (akaroml) [inline package.json]: debugging vm -> debuggee

File: package.json [hunk 11]
```diff
@@ -148,91 +148,78 @@
...
               },
-              "attachTimeout": {
+              "timeout": {
                 "type": "number",
                 "default": 30000,
-                "description": "The timeout when attaching to the debuggee VM"
+                "description": "Retry for this number of milliseconds to connect to remote debugging vm (Defaults to 30000 ms)."
```
Comment (akaroml) [inline package.json]: Timeout value before reconnecting, in milliseconds (default to 30000ms).

File: package.json [hunk 12]
```diff
@@ -148,91 +148,78 @@
...
                   "type": "string"
                 },
                 "default": [],
-                "description": "The source folders of the project."
+                "description": "The source directories of the program."
               },
               "projectName": {
                 "type": "string",
-                "description": "Name of the project that contains the startup class (Optional, only required if startupClass is contained in multiple projects)",
+                "description": "Name of the project that contains the main class (Optional, only required if the main class is contained in multiple projects).",
```
Comment (akaroml) [inline package.json]: "optional, only required when the main class resides in multiple projects"

File: package.json [hunk 13]
```diff
@@ -243,21 +230,22 @@
             "type": "java",
             "name": "Java Debug",
```
Comment (akaroml) [inline package.json]: Java Debug -> Debug (Launch)

File: package.json [hunk 14]
```diff
@@ -243,21 +230,22 @@
...
-            "options": []
+            "mainClass": "",
+            "args": "",
+            "sourcePaths": [
+              "${workspaceRoot}"
+            ]
           },
           {
             "type": "java",
             "name": "Java Debug (Attach)",
```
Comment (akaroml) [inline package.json]: Debug (Attach)

File: src/extension.ts [hunk 15]
```diff
@@ -81,26 +81,29 @@ export function activate(context: ExtensionContext) {
 								resolve();
 								commands.registerCommand(Commands.START_DEBUG_SESSION, async (config) => {
 									if (config.request === 'launch') {
-										if (!config.startupClass) {
-											vscode.window.showErrorMessage('Please specify startupClass in launch.json first.');
+										if (!config.mainClass) {
+											vscode.window.showErrorMessage('Please configure the main class in launch.json first.');
```
Comment (akaroml) [inline src/extension.ts]: configure -> specify

File: src/extension.ts [hunk 16]
```diff
@@ -81,26 +81,29 @@ export function activate(context: ExtensionContext) {
...
-												startupClass: config.startupClass,
+												mainClass: config.mainClass,
 												projectName: config.projectName,
 											};
-											const classpath = await languageClient.sendRequest(ClasspathResolveRequest.type, params);
-											config.classpath = classpath;
+											config.classPaths = await languageClient.sendRequest(ClasspathResolveRequest.type, params);
+										}
+										if (!config.classPaths || !Array.isArray(config.classPaths) || !config.classPaths.length) {
+											vscode.window.showErrorMessage('Cannot resolve the class paths automatically, please configure it in launch.json first.');
```
Comment (akaroml) [inline src/extension.ts]: class paths -> classpaths
configure it -> specify
and remove the "first"

File: src/extension.ts (hunk 15 above)
Comment (akaroml) [reply src/extension.ts]: and remove the "first"

File: src/extension.ts [hunk 17]
```diff
@@ -81,26 +81,29 @@ export function activate(context: ExtensionContext) {
...
+											return;
 										}
 									} else if (config.request === 'attach') {
-										if (!config.hostName || !config.port || !config.attachTimeout) {
-											vscode.window.showErrorMessage('Please specify hostName/port/attachTimeout in launch.json first.');
+										if (!config.hostName || !config.port) {
+											vscode.window.showErrorMessage('Please configure the host name and port of the attached VM in launch.json first.');
```
Comment (akaroml) [inline src/extension.ts]: configure -> specify
port of the attached VM -> the port of the remote debuggee
and remove "first"

**Review by akaroml — CHANGES_REQUESTED** [review]

**Review by akaroml — APPROVED** [review]
Thanks!

**Pull request merged**